spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,526 @@
+
+# %%
+df_valid[["text", "label"]].iloc[error_buckets[(1, 0)]].head()
+
+# %% [markdown]
+# We notice that we're mis-classifying particularly some comments with shortened urls (e.g. `bit.ly/...`) — these links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+
+# %% [markdown]
+# ## 3. Monitor data slices
+#
+# We leverage *slicing functions* (SFs) — an abstraction that shares syntax with *labeling functions*, which you should already be familiar with! (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).) A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
+
+# %% [markdown]
+# In the following cells, we define a slicing function that identifies these shortened links the spam dataset.
+# To do so, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined — SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing.sf import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Spam comments post links to other channels."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+slice_names = [sf.name for sf in sfs]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing.apply import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# With a utility function from `snorkel.slicing.monitor`, we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.slicing.monitor import PandasSlicer
+
+pd_slicer = PandasSlicer(df_valid, sfs, slice_names)
+short_link_df = pd_slicer.slice("short_link")
+short_link_df[["text", "label"]]
+
+# %% [markdown]
+# Now, we add labels for this particularly slice to an existing dataloader.
+# Specifically, `add_slice_labels` will add two sets of labels for each slice:
+# * `spam_task_slice:{slice_name}_ind`: an indicator label, which corresponds to the outputs of the slicing functions.
+# These indicate whether each example is in the slice (`label=1`)or not (`label=0`).
+# * `spam_task_slice:{slice_name}_pred`: a _masked_ set of the original task labels (in this case, labeled `spam_task`) for each slice. Examples that are masked (with `label=-1`) will not contribute to loss or scoring.
+
+# %%
+from snorkel.slicing.utils import add_slice_labels
+
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
```

Comment:
Similar to LFAnalysis, let's opt for passing in the list of sfs

Reply:
re: offline conversation (well, on Slack), we'll keep this for now and address it when we make the update to structured arrays (for named columns)